cmds/switch_root/switch_root.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+)
+
+var (
+	help    = flag.Bool("h", false, "Help")
+	version = flag.Bool("V", false, "Version")
+)
+
+// Return the usage string
+func usage() string {
+	return "switch_root [-h] [-V]\nswitch_root newroot init"
+}
+
+// Recursively deletes everything at slash
+// Does not continue down other filesystems i.e.
+// new_root, devtmpfs, profs and sysfs
+func littleDoctor(path string, fs *syscall.Statfs_t) error {
+	pathFS := syscall.Statfs_t{}
+
+	if err := syscall.Statfs(path, &pathFS); err != nil {
+		return err
+	}
+
+	if pathFS.Type != fs.Type {
+		return nil
+	}
+
+	file, err := os.Open(path)
+
+	if err != nil {
+		return fmt.Errorf("Could not open %s: %v", path, err)
+	}
+
+	if fileStat, _ := file.Stat(); fileStat.IsDir() {
+
+		names, err := file.Readdirnames(-1)
+		file.Close()
+
+		if err != nil {
+			return err
+		}
+
+		for _, fileName := range names {
+
+			if fileName == "." || fileName == ".." {
+				return nil
```

Comment:
You're right. I think the only reason it had been working is because "." and ".." are at the end of array of strings from file.Readditnames(-1). Should be continue.